day7/part2.go

```diff
@@ -0,0 +1,196 @@
+
+func Part2Opts(path string, cache CacheStrategy, proc ParallelStrategy) (int, error) {
+	if cache == Naive && proc != ForLoop {
+		return 0, fmt.Errorf("invalid strategy combination")
+	}
+
+	positions, err := utils.ReadIntCsv(path)
+	if err != nil {
+		return 0, err
+	}
+	maxPosition := utils.MaxInt(positions...)
+	solutions := make([]int, maxPosition+1)
+	costFunction := makeFuelStrategy(cache)
+	processor := makeProcessor(proc)
+
+	processor(positions, maxPosition, costFunction, solutions)
+
+	return utils.MinInt(solutions...), nil
+}
+
+func rawFuelCost(distance int) int {
+	/*
+		1 -> 1: 1
+		2 -> 3: 1 + 2
+		3 -> 6: 1 + 2 + 3
+		4 -> 10: 1 + 2 + 3 + 4
+
+		TODO: there's probably a cheaper / clever math trick to do this
+	*/
+	cost := 0
+
+	for i := 1; i <= distance; i++ {
+		cost += i
+	}
+	return cost
+}
+
+type CacheStrategy int
+
+const (
+	NoCache CacheStrategy = iota
+	Mutex
+	RWMutex
+	Naive
+)
+
+type costFunction = func(int) int
+
+func makeFuelStrategy(strategy CacheStrategy) costFunction {
+	if strategy == NoCache {
+		return rawFuelCost
+	}
+
+	fuelCostCache := make(map[int]int)
+
+	if strategy == Naive {
+		return func(distance int) int {
+			cached, ok := fuelCostCache[distance]
+			if ok {
+				return cached
+			}
+			cost := rawFuelCost(distance)
+			fuelCostCache[distance] = cost
+			return cost
+		}
+	}
+
+	if strategy == Mutex {
+		var mutex sync.Mutex
+		return func(distance int) int {
+			mutex.Lock()
+			defer mutex.Unlock()
+			cached, ok := fuelCostCache[distance]
+			if ok {
+				return cached
+			}
+			cost := rawFuelCost(distance)
+			fuelCostCache[distance] = cost
+			return cost
+		}
+	}
+
+	if strategy == RWMutex {
+		var rwMutex sync.RWMutex
+		return func(distance int) int {
+			rwMutex.RLock()
+			cached, ok := fuelCostCache[distance]
+			rwMutex.RUnlock()
+			if ok {
+				return cached
+			}
+
+			rwMutex.Lock()
+			defer rwMutex.Unlock()
+
+			cached, ok = fuelCostCache[distance]
+			if ok {
+				return cached
+			}
+
+			cost := rawFuelCost(distance)
+			fuelCostCache[distance] = cost
+			return cost
+		}
+	}
+	return nil
+}
+
+type ParallelStrategy int
+
+const (
+	ForLoop ParallelStrategy = iota
+	WaitGroup
+)
+
+type processor = func([]int, int, costFunction, []int)
+
+func makeProcessor(strategy ParallelStrategy) processor {
+	if strategy == ForLoop {
+		return forLoopProcessor
+	}
+	if strategy == WaitGroup {
+		return waitGroupProcessor
+	}
+	return nil
+}
+
+func forLoopProcessor(positions []int, maxPosition int, cost costFunction, solutions []int) {
+	for target := 0; target <= maxPosition; target++ {
+		fuel := 0
+		for _, position := range positions {
+			distance := utils.AbsInt(position - target)
+			fuel += cost(distance)
+		}
+		solutions[target] = fuel
+	}
+}
+
+func waitGroupProcessor(positions []int, maxPosition int, cost costFunction, solutions []int) {
+	var wg sync.WaitGroup
+	for target := 0; target <= maxPosition; target++ {
+		wg.Add(1)
+		go func(target int) {
+			defer wg.Done()
```

Comment:
WaitGroup is builtin so handy to use sometimes but it has an aweful interface "Don't forget to announce that the routine is done."  it says "Wait, don't _you_ know that?!" I say.

[Tomb](https://pkg.go.dev/gopkg.in/tomb.v2) is one that I like that doesn't seem very popular:
```golang
var t tomb.Tomb
t.Go(func() error {
   // stuff
   return nil
})
err := t.Wait()
```

[errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) seems more common, but lacks the ability to kill them all.